core/primitives/src/stateless_validation.rs

```diff
@@ -126,22 +128,59 @@ impl EncodedChunkStateWitness {
         const STATE_WITNESS_COMPRESSION_LEVEL: i32 = 3;
-        let borsh_bytes = borsh::to_vec(witness)?;
-        Ok((
-            Self(zstd::encode_all(borsh_bytes.as_slice(), STATE_WITNESS_COMPRESSION_LEVEL)?.into()),
-            borsh_bytes.len(),
-        ))
+
+        // Flow of data: State witness --> Borsh serialization --> Counting write --> zstd compression --> Bytes.
+        // CountingWrite will count the number of bytes for the Borsh-serialized witness, before compression.
+        let mut counting_write = CountingWrite::new(zstd::stream::Encoder::new(
+            Vec::new().writer(),
+            STATE_WITNESS_COMPRESSION_LEVEL,
+        )?);
+        borsh::to_writer(&mut counting_write, witness)?;
+
+        let borsh_bytes_len = counting_write.bytes_written();
+        let encoded_bytes = counting_write.into_inner().finish()?.into_inner();
+
+        Ok((Self(encoded_bytes.into()), borsh_bytes_len.as_u64() as usize))
     }
 
     /// Decompress and borsh-deserialize encoded witness bytes.
     /// Returns decoded witness along with the raw (uncompressed) witness size.
     pub fn decode(&self) -> std::io::Result<(ChunkStateWitness, ChunkStateWitnessSize)> {
         // We want to limit the size of decompressed data to address "Zip bomb" attack.
         // The value here is the same as NETWORK_MESSAGE_MAX_SIZE_BYTES.
-        const MAX_WITNESS_SIZE: usize = 512 * bytesize::MIB as usize;
-        let borsh_bytes = decompress_with_limit(self.0.as_ref(), MAX_WITNESS_SIZE)?;
-        let witness = ChunkStateWitness::try_from_slice(&borsh_bytes)?;
-        Ok((witness, borsh_bytes.len()))
+        const MAX_WITNESS_SIZE: ByteSize = ByteSize::mib(512);
+
+        self.decode_with_limit(MAX_WITNESS_SIZE)
+    }
+
+    /// Decompress and borsh-deserialize encoded witness bytes.
+    /// Returns decoded witness along with the raw (uncompressed) witness size.
+    pub fn decode_with_limit(
+        &self,
+        limit: ByteSize,
+    ) -> std::io::Result<(ChunkStateWitness, ChunkStateWitnessSize)> {
+        // Flow of data: Bytes --> zstd decompression --> Counting read --> Borsh deserialization --> State witness.
+        // CountingRead will count the number of bytes for the Borsh-deserialized witness, after decompression.
+        let mut counting_read = CountingRead::new_with_limit(
+            zstd::stream::Decoder::new(self.0.as_ref().reader())?,
+            limit,
+        );
+
+        match borsh::from_reader(&mut counting_read) {
+            Err(err) => {
+                // If decompressed data exceeds the limit then CountingRead will return a WriteZero error.
+                // Here we convert it to a more descriptive error to make debugging easier.
+                let err = if err.kind() == std::io::ErrorKind::WriteZero {
+                    std::io::Error::other(format!(
+                        "Decompressed data exceeded limit of {limit} bytes: {err}"
```

Comment:
nit: `ByteSize` display implementation adds units, so no need for `bytes` here:
`"Decompressed data exceeded limit of {limit}: {err}"`

Reply:
Done.